platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java

```diff
@@ -265,535 +292,130 @@ public Hash hash(
-        // reserved segment).
-        long lastPath = -1;
-        while (itr.hasNext()) {
-            final VirtualLeafRecord<K, V> next = itr.peek();
-            final long path = next.getPath();
-            final int rank = getRank(path);
-
-            // SANITY CHECK: Fail fast if this condition does not hold
-            if (path < lastPath) {
-                throw new IllegalStateException("The paths in the iterator must be strictly increasing! " + "lastPath="
-                        + lastPath + ", path=" + path);
-            }
-            lastPath = path;
-
-            // The path must always be within this range. We can use an assertion here because,
-            // unless there is some bug in our code, this cannot happen.
-            assert path >= firstLeafPath && path <= lastLeafPath
-                    : "Invalid path lies outside the leaf path range " + path;
-
-            // Break out of the loop if the next leaf to process is in the reserved space
-            if (rank == maxLeafRank && path >= lastLeafPath - reservedLastLeafCount) {
-                break;
-            }
-
-            // Depending on the rank we're processing, we need to know the segment size.
-            final long segmentSize = rank == minLeafRank ? minRankSegmentSize : maxRankSegmentSize;
-            // Either eof is the end of the rank, or the last leaf before the reserved section
-            final long eofPath = Math.min((1L << (rank + 1)) - 1, lastLeafPath - reservedLastLeafCount + 1);
-
-            // Populate the wq with leaves by accumulating them from the iterator, a segment at a time.
-            final HashingQueue<K, V> wq = queue1.reset();
-            accumulate(itr, wq, path - (getIndexInRank(path) % segmentSize), segmentSize, eofPath);
-
-            // Setup and hash the subtree that we have accumulated. If I am hashing the minLeafRank,
-            // then the results go into the minRankStopQueue. If I am hashing the maxLeafRank, then
-            // the results go into the maxRankStopQueue.
-            final HashingQueue<K, V> pq = queue2.reset();
-            final HashingQueue<K, V> sq = rank == maxLeafRank ? maxRankStopQueue : minRankStopQueue;
-            listener.onBatchStarted();
-            hashSubTree(hashReader, listener, wq, pq, null, sq, firstLeafPath, lastLeafPath, rank, stopRank);
-            listener.onBatchCompleted();
+        final HashMap<Long, ChunkHashTask> map = new HashMap<>();
+        ChunkHashTask resultTask = new ChunkHashTask(hashingPool, INVALID_PATH, 1);
+        int rootTaskHeight = Math.min(firstLeafRank, chunkHeight);
+        ChunkHashTask rootTask = new ChunkHashTask(hashingPool, ROOT_PATH, rootTaskHeight);
+        rootTask.setOut(resultTask);
+        map.put(ROOT_PATH, rootTask);
+
+        boolean firstLeaf = true;
+        final long[] stack = new long[lastLeafRank + 1];
+        Arrays.fill(stack, INVALID_PATH);
+
+        final int[] parentRankHeights = new int[256];
```

Comment:
Minor: not sure where this magic number comes from. Would be nice to either have a comment here in code, or put the number in a well-named constant.

Reply:
OK, I'll add a short comment